hw/drivers/uart/uart_da1469x/src/uart_da1469x.c

```diff
@@ -0,0 +1,563 @@
+    switch (uc->uc_parity) {
+    case UART_PARITY_NONE:
+        break;
+    case UART_PARITY_EVEN:
+        reg |= UART2_UART2_LCR_REG_UART_EPS_Msk;
+        /* no break */
+    case UART_PARITY_ODD:
+        reg |= UART2_UART2_LCR_REG_UART_PEN_Msk;
+        break;
+    }
+    reg |= (uc->uc_stopbits - 1) << UART2_UART2_LCR_REG_UART_STOP_Pos;
+    reg |= (uc->uc_databits - 5) << UART2_UART2_LCR_REG_UART_DLS_Pos;
+    uart->UART2_LCR_REG = reg;
+
+    /* Enable flow-control if requested and supported */
+    if (uc->uc_flow_ctl == UART_FLOW_CTL_RTS_CTS) {
+        uart->UART2_MCR_REG |= UART2_UART2_MCR_REG_UART_AFCE_Msk |
+                               UART2_UART2_MCR_REG_UART_RTS_Msk;
+    }
+    /* Enable hardware FIFO */
+    uart->UART2_SFE_REG = UART2_UART2_SFE_REG_UART_SHADOW_FIFO_ENABLE_Msk;
+    uart->UART2_SRT_REG = 0;
+    uart->UART2_STET_REG = 0;
+
+    dev->active = 1;
+    dev->rx_stalled = 0;
+    dev->tx_started = 0;
+
+    da1469x_uart_lsr_intr_enable(dev);
+
+    mcu_gpio_set_pin_function(dev->da1469x_cfg.pin_tx, MCU_GPIO_MODE_OUTPUT,
+                              dev->hw->tx_pin_func);
+
+    if (dev->da1469x_cfg.pin_rx >= 0) {
+        mcu_gpio_set_pin_function(dev->da1469x_cfg.pin_rx,
+                                  dev->da1469x_cfg.rx_pullup ? MCU_GPIO_MODE_INPUT_PULLUP : MCU_GPIO_MODE_INPUT,
+                                  dev->hw->rx_pin_func);
+        da1469x_uart_rx_intr_enable(dev);
+    }
+
+    if (dev->da1469x_cfg.pin_rts >= 0) {
+        mcu_gpio_set_pin_function(dev->da1469x_cfg.pin_rts, MCU_GPIO_MODE_OUTPUT,
+                                  dev->hw->rts_pin_func);
+    }
+
+    if (dev->da1469x_cfg.pin_cts >= 0) {
+        mcu_gpio_set_pin_function(dev->da1469x_cfg.pin_cts, MCU_GPIO_MODE_INPUT,
+                                  dev->hw->cts_pin_func);
+    }
+
+    NVIC_ClearPendingIRQ(irqn);
+    NVIC_EnableIRQ(irqn);
+}
+
+static void
+da1469x_uart_uart_shutdown(struct da1469x_uart_dev *dev)
+{
+    dev->active = 0;
+
+    /*
+     * Reset UART before disabling clock to avoid any unexpected behavior after
+     * clock is enabled again.
+     */
+    dev->hw->regs->UART2_SRR_REG = UART2_UART2_SRR_REG_UART_UR_Msk |
+                                   UART2_UART2_SRR_REG_UART_RFR_Msk |
+                                   UART2_UART2_SRR_REG_UART_XFR_Msk;
+    NVIC_DisableIRQ(dev->hw->irqn);
+    NVIC_ClearPendingIRQ(dev->hw->irqn);
+
+    CRG_COM->RESET_CLK_COM_REG = dev->hw->clk_com_mask;
+
+    da1469x_pd_release(MCU_PD_DOMAIN_COM);
+}
+
+static int
+da1469x_uart_open(struct os_dev *odev, uint32_t wait, void *arg)
+{
+    struct da1469x_uart_dev *dev = (struct da1469x_uart_dev *)odev;
+    struct uart_conf *uc = (struct uart_conf *)arg;
+    (void)wait;
+
+    if (!uc) {
+        return OS_EINVAL;
+    }
+
+    if (odev->od_flags & OS_DEV_F_STATUS_OPEN) {
+        return OS_EBUSY;
+    }
+
+    dev->uc = *uc;
+
+    if (uc->uc_speed < 1200 || uc->uc_speed > 1000000) {
+        return SYS_ENOTSUP;
```

Comment:
use OS error codes, as above, since this is os_* call

Reply:
fixed